integration/testutil.go

```diff
@@ -101,37 +101,148 @@ level = "debug"
+  init: true
+  entrypoint: [ "sleep", "infinity" ]
+  environment:
+   - NO_PROXY=127.0.0.1,localhost,{{.RegistryHost}}:443
+  tmpfs:
+   - /tmp:exec,mode=777
+   - /var/lib/containerd
+   - /var/lib/soci-snapshotter-grpc
+  volumes:
+   - /dev/fuse:/dev/fuse
+ registry:
+  image: ghcr.io/oci-playground/registry:v3.0.0-alpha.1
+  container_name: {{.RegistryHost}}
+  environment:
+   - REGISTRY_AUTH=htpasswd
+   - REGISTRY_AUTH_HTPASSWD_REALM="Registry Realm"
+   - REGISTRY_AUTH_HTPASSWD_PATH=/auth/htpasswd
+   - REGISTRY_HTTP_TLS_CERTIFICATE=/auth/domain.crt
+   - REGISTRY_HTTP_TLS_KEY=/auth/domain.key
+   - REGISTRY_HTTP_ADDR={{.RegistryHost}}:443
+  volumes:
+   - {{.AuthDir}}:/auth:ro
+{{.NetworkConfig}}
+`
+const composeRegistryAltTemplate = `
+version: "3.7"
+services:
+  {{.ServiceName}}:
+    image: soci_integ_test
+    privileged: true
+    init: true
+    entrypoint: [ "sleep", "infinity" ]
+    environment:
+    - NO_PROXY=127.0.0.1,localhost,{{.RegistryHost}}:443
+    tmpfs:
+    - /tmp:exec,mode=777
+    - /var/lib/containerd
+    - /var/lib/soci-snapshotter-grpc
+    volumes:
+    - /dev/fuse:/dev/fuse
+  registry:
+    image: ghcr.io/oci-playground/registry:v3.0.0-alpha.1
+    container_name: {{.RegistryHost}}
+    environment:
+    - REGISTRY_AUTH=htpasswd
+    - REGISTRY_AUTH_HTPASSWD_REALM="Registry Realm"
+    - REGISTRY_AUTH_HTPASSWD_PATH=/auth/htpasswd
+    - REGISTRY_HTTP_TLS_CERTIFICATE=/auth/domain.crt
+    - REGISTRY_HTTP_TLS_KEY=/auth/domain.key
+    - REGISTRY_HTTP_ADDR={{.RegistryHost}}:443
+    volumes:
+    - {{.AuthDir}}:/auth:ro
+  registry-alt:
+    image: registry:2
+    container_name: {{.RegistryAltHost}}
+`
+
+const composeBuildTemplate = `
+version: "3.7"
+services:
+ {{.ServiceName}}:
+  image: soci_integ_test
+  build:
+   context: {{.ImageContextDir}}
+   target: {{.TargetStage}}
+   args:
+    - SNAPSHOTTER_BUILD_FLAGS="-race"
+ registry:
+  image: ghcr.io/oci-playground/registry:v3.0.0-alpha.1
+ registry-alt:
+  image: registry:2
+`
+
+type dockerComposeYaml struct {
+	ServiceName     string
+	ImageContextDir string
+	TargetStage     string
+	RegistryHost    string
+	RegistryAltHost string
+	AuthDir         string
+	NetworkConfig   string
+}
 
 // getContainerdConfigYaml creates a containerd config yaml, by appending all
 // `additionalConfigs` to the default `containerdConfigTemplate`.
 func getContainerdConfigYaml(t *testing.T, disableVerification bool, additionalConfigs ...string) []byte {
 	if !isTestingBuiltinSnapshotter() {
 		additionalConfigs = append(additionalConfigs, proxySnapshotterConfig)
 	}
-
-	return []byte(testutil.ApplyTextTemplate(t, containerdConfigTemplate, struct {
+	s, err := testutil.ApplyTextTemplate(containerdConfigTemplate, struct {
 		DisableVerification bool
 		AdditionalConfig    string
 	}{
 		DisableVerification: disableVerification,
 		AdditionalConfig:    strings.Join(additionalConfigs, "\n"),
-	}))
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	return []byte(s)
```

Comment:
Merge conflict with #378 if that lands first. This is not a blocker, just a heads up; I can fix on my end if this lands first.